src/net.cpp

```diff
@@ -1312,46 +1298,150 @@ void CConnman::SocketHandler()
             if (pnode->hSocket == INVALID_SOCKET)
                 continue;
 
-            FD_SET(pnode->hSocket, &fdsetError);
-            hSocketMax = std::max(hSocketMax, pnode->hSocket);
-            have_fds = true;
-
+            error_set.insert(pnode->hSocket);
             if (select_send) {
-                FD_SET(pnode->hSocket, &fdsetSend);
+                send_set.insert(pnode->hSocket);
                 continue;
             }
             if (select_recv) {
-                FD_SET(pnode->hSocket, &fdsetRecv);
+                recv_set.insert(pnode->hSocket);
             }
         }
     }
 
-    int nSelect = select(have_fds ? hSocketMax + 1 : 0,
-                         &fdsetRecv, &fdsetSend, &fdsetError, &timeout);
+    return !recv_set.empty() || !send_set.empty() || !error_set.empty();
+}
+
+#ifdef USE_POLL
+void CConnman::SocketEvents(std::set<SOCKET> &recv_set, std::set<SOCKET> &send_set, std::set<SOCKET> &error_set)
+{
+    std::set<SOCKET> recv_select_set, send_select_set, error_select_set;
+    if (!GenerateSelectSet(recv_select_set, send_select_set, error_select_set)) {
+        interruptNet.sleep_for(std::chrono::milliseconds(SELECT_TIMEOUT_MILLISECONDS));
+        return;
+    }
+
+    std::map<SOCKET, struct pollfd> pollfds;
+    for (SOCKET hSocket : recv_select_set) {
+        pollfds[hSocket].fd = hSocket;
+        pollfds[hSocket].events |= POLLIN;
+    }
+
+    for (SOCKET hSocket : send_select_set) {
+        pollfds[hSocket].fd = hSocket;
+        pollfds[hSocket].events |= POLLOUT;
+    }
+
+    for (SOCKET hSocket : error_select_set) {
+        pollfds[hSocket].fd = hSocket;
+        // These flags are ignored, but we set them for clarity
+        pollfds[hSocket].events |= POLLERR|POLLHUP;
+    }
+
+    std::vector<struct pollfd> vpollfds;
+    vpollfds.reserve(pollfds.size());
+    for(auto it : pollfds)
+        vpollfds.push_back(std::move(it.second));
+
+    if(poll(vpollfds.data(), vpollfds.size(), SELECT_TIMEOUT_MILLISECONDS) < 0)
+        return;
+
+    if (interruptNet)
+        return;
+
+    for (struct pollfd pollfd : vpollfds) {
+        if (pollfd.revents & POLLIN)
```

Comment:
Style nit: braces whenever the next lines are indented (or put the then clause on the same line as the `if`).